### PR TITLE
MM-10833: link other channel references

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
Addresses missing channel references in some system messages, and in the channel header by relying on a new, computed props sent down with channels from the server.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-10833

#### Checklist
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Added or updated unit tests (required for all new features)
- [x] Has server changes: https://github.com/mattermost/mattermost-server/pull/8953

### CODE REVIEW

**Comment by lieut-data** [comment]
Rebased against `master` in hopes that unexpected build errors go away -- they aren't happening locally.

File: components/channel_header/channel_header.jsx [hunk 1]
```diff
@@ -317,7 +318,7 @@ export default class ChannelHeader extends React.Component {
             >
                 <Markdown
                     message={channel.header}
-                    options={headerMarkdownOptions}
+                    options={{...headerMarkdownOptions, channelNamesMap}}
```
Comment (hmhealey) [inline components/channel_header/channel_header.jsx]: Creating a new object for options on each render is going to cause the Markdown component to also re-render. I don't know how often ChannelHeader re-renders, but we might want to memoize this if that happens a lot.

File: components/channel_header/channel_header.jsx [hunk 2]
```diff
@@ -853,7 +854,7 @@ export default class ChannelHeader extends React.Component {
                         <span onClick={Utils.handleFormattedTextClick}>
                             <Markdown
                                 message={channel.header}
-                                options={headerMarkdownOptions}
+                                options={{...headerMarkdownOptions, channelNamesMap}}
```
Comment (hmhealey) [inline components/channel_header/channel_header.jsx]: Same here

File: components/channel_info_modal/channel_info_modal.jsx [hunk 3]
```diff
@@ -77,7 +79,12 @@ export default class ChannelInfoModal extends React.PureComponent {
 
         let channelPurpose;
         if (channel.purpose) {
-            channelPurpose = channel.purpose;
+            channelPurpose = (
+                <Markdown
+                    message={channel.purpose}
+                    options={{...headerMarkdownOptions, channelNamesMap}}
```
Comment (hmhealey) [inline components/channel_info_modal/channel_info_modal.jsx]: And here

File: components/channel_info_modal/channel_info_modal.jsx [hunk 4]
```diff
@@ -115,7 +122,7 @@ export default class ChannelInfoModal extends React.PureComponent {
                     <div className='info__value'>
                         <Markdown
                             message={channel.header}
-                            options={headerMarkdownOptions}
+                            options={{...headerMarkdownOptions, channelNamesMap}}
```
Comment (hmhealey) [inline components/channel_info_modal/channel_info_modal.jsx]: And finally here

File: components/edit_channel_purpose_modal/edit_channel_purpose_modal.jsx [hunk 5]
```diff
@@ -114,13 +87,13 @@ export default class EditChannelPurposeModal extends React.Component {
...
     handleSave = () => {
-        const {channel, actions: {patchChannel}} = this.props;
+        const {channel, actions: {getChannel, patchChannel}} = this.props;
         const {purpose} = this.state;
         if (!channel) {
             return;
         }
 
-        patchChannel(channel.id, {purpose});
+        patchChannel(channel.id, {purpose}).then(() => getChannel(channel.id));
```
Comment (hmhealey) [inline components/edit_channel_purpose_modal/edit_channel_purpose_modal.jsx]: Do we still need to get the channel after patching something that isn't the header? Shouldn't the client not replace the previous props after a patch?

If we need to do this all the time when we patch the channel, we may want to change the API to always return the channel with the new props set

File: components/post_markdown/system_message_helpers.jsx [hunk 6]
```diff
@@ -243,9 +244,13 @@ function renderPurposeChangeMessage(post) {
...
     }
 
+    const purposeOptions = {
+        channelNamesMap: post.props && post.props.channel_mentions,
+    };
+
     const username = renderUsernameForUserIdAndUsername(post.user_id, post.props.username);
-    const oldPurpose = post.props.old_purpose;
-    const newPurpose = post.props.new_purpose;
+    const oldPurpose = post.props.old_purpose ? renderFormattedText(post.props.old_purpose, purposeOptions) : null;
```
Comment (hmhealey) [inline components/post_markdown/system_message_helpers.jsx]: Why are we formatting the purpose here? The main place that's used is in the More Channels list where we don't format it

Comment (lieut-data) [reply components/post_markdown/system_message_helpers.jsx]: I had run across this in actual system messages, and thought that the purpose should be treated similarly to the header. Should the purpose remain plain text instead?

image

File: components/edit_channel_purpose_modal/edit_channel_purpose_modal.jsx (hunk 5 above)
Comment (lieut-data) [reply components/edit_channel_purpose_modal/edit_channel_purpose_modal.jsx]: See below, but I had been treating the header and purpose as both formattable fields.

You're right: we should be returning the updated channel in the patch itself. In fact, things are already wired up to do that, but I failed to make the patch endpoint actually fill in the channel props. I'll fix that!

File: components/post_markdown/system_message_helpers.jsx (hunk 6 above)
Comment (hmhealey) [reply components/post_markdown/system_message_helpers.jsx]: Yeah, the purpose is just plain text. We don't format it when it appears in the channel info modal or the more channels list

Comment (lieut-data) [reply components/post_markdown/system_message_helpers.jsx]: Sounds good -- I've reverted this part of the change.

**Comment by hmhealey** [comment]
Waiting on the server PR: https://github.com/mattermost/mattermost-server/pull/8953

**Pull request merged**